plugins/modules/keycloak_authz_permission.py

```diff
@@ -0,0 +1,471 @@
+        for scope in scopes:
+            s = kc.get_authz_authorization_scope_by_name(scope, cid, realm)
+            if r and not s['id'] in resource_scopes:
+                module.fail_json(msg='Resource %s does not include scope %s for client %s in realm %s' % (resources[0], scope, client_id, realm))
+            else:
+                payload['scopes'].append(s['id'])
+
+    elif permission_type == 'resource':
+        if resources:
+            for resource in resources:
+                r = kc.get_authz_resource_by_name(resource, cid, realm)
+                if not r:
+                    module.fail_json(msg='Unable to find authorization resource with name %s for client %s in realm %s' % (resource, cid, realm))
+                else:
+                    payload['resources'].append(r['_id'])
+
+    # Add policy ids, if any, to the payload.
+    if policies:
+        for policy in policies:
+            p = kc.get_authz_policy_by_name(policy, cid, realm)
+
+            if p:
+                payload['policies'].append(p['id'])
+            else:
+                module.fail_json(msg='Unable to find authorization policy with name %s for client %s in realm %s' % (policy, client_id, realm))
+
+    # Add "id" to payload for update operations
+    if permission:
+        payload['id'] = permission['id']
+
+        # Handle the special case where the user attempts to change an already
+        # existing permission's type - something that can't be done without a
+        # full delete -> (re)create cycle.
+        if permission['type'] != payload['type']:
+            module.fail_json(msg='Modifying the type of permission (scope/resource) is not supported: \
+                                  permission %s of client %s in realm %s unchanged' % (permission['id'], cid, realm))
+
+    # Updating an authorization  permission is tricky for several reasons.
+    # Firstly, the current permission is retrieved using a _policy_ endpoint,
+    # not from a permission endpoint. Also, the data that is returned is in a
+    # different format than what is expected by the payload. So, comparing the
+    # current state attribute by attribute to the payload is not possible.  For
+    # example the data contains a JSON object "config" which may contain the
+    # authorization type, but which is no required in the payload.  Moreover,
+    # information about resources, scopes and policies is _not_ present in the
+    # data. So, there is no way to determine if any of those fields have
+    # changed. Therefore the best options we have are
+    #
+    # a) Always apply the payload without checking the current state
+    # b) Refuse to make any changes to any settings (only support create and delete)
+    #
+    # The approach taken here is a).
+    #
+    if permission and state == 'present':
+        kc.update_authz_permission(payload=payload, permission_type=permission_type, id=permission['id'], client_id=cid, realm=realm)
+        result['changed'] = False
```

Comment:
If anything is updated, `changed` must be set to `True`. So either the above function call isn't needed, or the assignment to `False` is wrong.

Reply:
Let me explain this - this is also tangentially related to check_mode. Keycloak authorization permission are handled "interestingly". When you create or remove a permission, you use a "permission endpoint" or URL. When you query existing permissions, you actually query the "policies endpoint". To make matters worse, these two endpoints return or expect different type of data:

* Data from the *policies* endpoint does not include information about resources, policies or scopes (all lists)
* Data from *policies* endpoint is returned in a different format than what the *permission* endpoints expect when creating authorization permissions

The end result is that it is not possible to detect the current state of the permission. This means there are two options:

1. Only support "create" and "delete". Ignore any attempts to manage the desired state of created objects.
1. Always update all objects that already exists. This maintains the desired state but it is impossible to know if the update was actually needed or not.
 
I would prefer option 2) as there's no real harm in doing it afaics. Whether *result['changed']* should be False or True or None or what I don't know. That more about Ansible conventions I guess than anything else.

Option 1) would be, in my opinion, more surprising and limit the usability of this module.

Thoughts?

Reply:
Option 2) is fine for me, but this needs to be documented, and `changed=True` should always be returned when you aren't sure whether something was changed.